Applications/Spire/Include/Spire/Ui/ListView.hpp

```diff
@@ -0,0 +1,222 @@
+       * @param submission The submitted value.
+       */
+      using SubmitSignal = Signal<void (const QString& submission)>;
+  
+      /**
+       * Constructs a ListView.
+       * @param current_model The current value's model.
+       * @param list_model The list model which holds a list of items.
+       * @param factory A function that takes a ListMoel and a index
+       *                used to construct a widget displayed in the ListView.
+       * @param parent The parent widget.
+       */
+      ListView(std::shared_ptr<CurrentModel> current_model,
+        std::shared_ptr<ArrayListModel> list_model,
+        std::function<QWidget* (
+          std::shared_ptr<ArrayListModel>, int index)> factory,
+        QWidget* parent = nullptr);
+  
+      /** Returns the current model. */
+      const std::shared_ptr<CurrentModel>& get_current_model() const;
+
+      /** Returns the list model. */
+      const std::shared_ptr<ArrayListModel>& get_list_model() const;
+
+      /** Returns the direction of the ListView. */
+      Qt::Orientation get_direction() const;
+
+      /** Sets the direction of the ListView. */
+      void set_direction(Qt::Orientation direction);
+      
+      /** Returns the navigation behavior of the ListView. */
+      EdgeNavigation get_edge_navigation() const;
+
+      /**
+       * Sets the navigation behavior of the ListView.
+       * @param navigation The keyboard navigation behavior when the first or
+       *                    last list item is selected.
+       */
+      void set_edge_navigation(EdgeNavigation navigation);
+
+      /** Returns the overflow mode of the ListView. */
+      Overflow get_overflow() const;
+
+      /**
+       * Sets the overflow mode of the ListView.
+       * @param overflow The overflow model.
+       */
+      void set_overflow(Overflow overflow);
+
+      /** Returns the selection mode of the ListView. */
+      SelectionMode get_selection_mode() const;
+
+      /**
+       * Sets the selection mode of the ListView.
+       * @param selection_mode The selection mode for the ListView.
+       */
+      void set_selection_mode(SelectionMode selection_mode);
+
+      /**
+       * Returns whether items are selected when focused for
+       * selection_mode = SINGLE.
+       */
+      bool get_selection_follows_focus() const;
+
+      /**
+       * Sets whether items are selected when focused for
+       * selection_mode = SINGLE.
+       * @param is_selection_follows_focus True iff items are selected
+       *                                   when focused
+       */
+      void set_selection_follows_focus(bool is_selection_follows_focus);
+
+      /** Returns the value of the selected list item. */
+      const QString& get_selected() const;
+
+      /**
+       * Returns the ListItem connected the specified text value.
+       * @param value The text value associated with an item.
+       */
+      const ListItem* get_item(const QString& value) const;
```

Comment:
Items can be accessed because it is placed in Data section according to the specs. Otherwise, the ScollableListBox needs to get the position of the current ListItem to adjust the scrollbar's position. 

The ListView is supposed to provide the method to get a list of ListItems based on the specs, but the list of ListItems are variable which might be removed or added, so I provide this method to obtain a ListItem by the value rather than all ListItems.

Reply:
When you implement the ```ScrollableListBox``` we can revisit whether this method is needed, if it is then you can add it as part of that task. I am very hesitant to add things to the API that don't have an immediate need, because everything in the API needs to be supported forever.